src/NHibernate/Hql/Ast/ANTLR/Tree/BetweenOperatorNode.cs

```diff
@@ -63,25 +63,27 @@ private IASTNode GetHighOperand()
 
 		private static void Check(IASTNode check, IASTNode first, IASTNode second)
 		{
-			var expectedTypeAwareNode = check as IExpectedTypeAwareNode;
-			if (expectedTypeAwareNode != null)
+			if (!(check is IExpectedTypeAwareNode expectedTypeAwareNode) ||
+			    expectedTypeAwareNode.ExpectedType != null)
 			{
-				IType expectedType = null;
-				var firstNode = first as SqlNode;
-				if (firstNode != null)
-				{
-					expectedType = firstNode.DataType;
-				}
-				if (expectedType == null)
+				return;
+			}
+
+			IType expectedType = null;
+			if (first is SqlNode firstNode)
+			{
+				expectedType = firstNode.DataType;
+			}
+
+			if (expectedType == null)
+			{
+				if (second is SqlNode secondNode)
 				{
-					var secondNode = second as SqlNode;
-					if (secondNode != null)
-					{
-						expectedType = secondNode.DataType;
-					}
+					expectedType = secondNode.DataType;
 				}
-				expectedTypeAwareNode.ExpectedType = expectedType;
 			}
+
+			expectedTypeAwareNode.ExpectedType = expectedType;
```

Comment:
This is probably readable:

```suggestion
			expectedTypeAwareNode.ExpectedType = (first as SqlNode)?.DataType ?? (second as SqlNode)?.DataType;
```

Reply:
Changed, as it is readable to me. I couldn't commit the suggestion due to `Applying suggestions on deleted lines is not supported`.